### ISSUE TITLE
is there a function that raise an exception when coerce fail?

### ISSUE THREAD

**Issue by DogLooksGood** [wilkerlucio/spec-coerce#6]
Personally I think it's a common usage to do coerce and assert together. what I'm looking for is a function do the following but won't walk the data structure twice:
```clojure
(s/def ::number int?)

(defn the-fn [k x]
  (->> (sc/coerce k x)
       (s/assert* k)))

(the-fn ::number 10)     ;; => 10
(the-fn ::number "10")   ;; => 10
(the-fn ::number "10.1") ;; => exception
```

**Comment by wilkerlucio** [comment]
Nothing by default, but it seems easy to create one, do you need anything more than what this example function is providing?

**Comment by DogLooksGood** [comment]
nothing more i think. i'm just confused with the current coerce behavior, why there's no error or return a result represent invalid when the data can't be coerce? in this case, after coerce i still don't know if the data is valid or not.

**Comment by carocad** [comment]
I agree with @DogLooksGood. The current behavior of Clojure spec is to return ::invalid on a failed conform.

Considering that conform and conformers are the current (thought incorrect) way of handling coercion I think this would be an expected behavior.

**Comment by wilkerlucio** [comment]
@carocad 
The idea to keep things is just to don't break the structure itself, the `coerce` method is for a single item coercion, this API I don't think on changing, but I can provide a new one like `coerce!` that throws an error when the coercion fails, what you think?

**Comment by wilkerlucio** [comment]
@carocad @DogLooksGood 

Hello guys, I was thinking about it, I have this idea for the fns, would like your input on that:

* `sc/coerce` - will keep working as is, returns original value
* `sc/coerce!` - throws exception when coercion fails
* `sc/conform` - returns value coerced and conformed, returning `::s/invalid` as the conform does when doesn't work

And question, how are you using it? Are you hitting `sc/coerce` directly using map specs or other things, or using the `coerce-structure` for maps?

**Comment by DogLooksGood** [comment]
I would use it for apis, so i care about both validation and type transform.

I think it's not good to have a function named conform, then people have s/conform and sc/conform. from the viewpoint of a user like me, i think it's kind of messy.

**Comment by carocad** [comment]
I think that API is fine. Although I understand the point of dog looks good, I am not able to come up with a more appropriate name.

I think that namespaces are meant to precisely cover the case of functions having the same name.

> And question, how are you using it?

I'm using coerce directly with map Spec. To be honest I didn't understand why there was a second function for coercion :)

**Comment by wilkerlucio** [comment]
Just released on `1.0.0-alpha7`
